File: python-mvp/engine/propagator.py
"""
Propagator — FR-5005 boundary. Discrete regime taxonomy + maneuver cost table, mirroring
server/src/engine/Propagator.ts (FS-104's Maneuver Cost Table). Sprint-0 fix: TypeScript's
Propagator.advance() had zero production callers (orbital math never ran, CLAUDE.md's documented
defect) — this port's GameEngine calls advance() from the same turn-end hook that ticks deploy
states and maneuvers, so it always runs.
"""
from __future__ import annotations

from .types import Asset, OrbitalRegimeLabel

MEAN_MOTION_DEG_PER_TICK = {"LEO": 24, "MEO": 6, "GEO": 1.5}

ALTITUDE_COST = {
    "LEO-LEO": (0, 0), "MEO-MEO": (0, 0), "GEO-GEO": (0, 0),
    "LEO-MEO": (3, 2), "MEO-LEO": (3, 2),
    "MEO-GEO": (1, 3), "GEO-MEO": (1, 3),
    "LEO-GEO": (4, 4), "GEO-LEO": (4, 4),
}

PLANE_COST = {
    "LEO": {
        "EQUATORIAL-EQUATORIAL": (0, 0), "PROGRADE-PROGRADE": (0, 0), "POLAR-POLAR": (0, 0),
        "EQUATORIAL-PROGRADE": (6, 2), "PROGRADE-EQUATORIAL": (6, 2),
        "PROGRADE-POLAR": (6, 2), "POLAR-PROGRADE": (6, 2),
        "EQUATORIAL-POLAR": (11, 3), "POLAR-EQUATORIAL": (11, 3),
    },
    "MEO": {
        "EQUATORIAL-EQUATORIAL": (0, 0), "PROGRADE-PROGRADE": (0, 0), "POLAR-POLAR": (0, 0),
        "EQUATORIAL-PROGRADE": (3, 1), "PROGRADE-EQUATORIAL": (3, 1),
        "PROGRADE-POLAR": (3, 1), "POLAR-PROGRADE": (3, 1),
        "EQUATORIAL-POLAR": (5, 2), "POLAR-EQUATORIAL": (5, 2),
    },
    "GEO": {
        "EQUATORIAL-EQUATORIAL": (0, 0), "PROGRADE-PROGRADE": (0, 0), "POLAR-POLAR": (0, 0),
        "EQUATORIAL-PROGRADE": (2, 1), "PROGRADE-EQUATORIAL": (2, 1),
        "PROGRADE-POLAR": (2, 1), "POLAR-PROGRADE": (2, 1),
        "EQUATORIAL-POLAR": (4, 1), "POLAR-EQUATORIAL": (4, 1),
    },
}

COMBINED_MANEUVER_DISCOUNT = 0.25


def _parse(label: OrbitalRegimeLabel) -> tuple[str, str]:
    altitude, plane = label.split("-")
    return altitude, plane


class Propagator:
    def __init__(self) -> None:
        self._mean_anomaly: dict[str, float] = {}

    def advance(self, assets: list[Asset]) -> None:
        for asset in assets:
            altitude, _ = _parse(asset.true_regime)
            current = self._mean_anomaly.get(asset.asset_id, 0.0)
            self._mean_anomaly[asset.asset_id] = (current + MEAN_MOTION_DEG_PER_TICK[altitude]) % 360

    def current_regime(self, asset: Asset) -> OrbitalRegimeLabel:
        return asset.true_regime

    def plan_maneuver(self, asset: Asset, target_regime: OrbitalRegimeLabel) -> dict:
        if asset.maneuver_state is not None:
            return {"turns_required": 0, "fuel_cost": 0, "rejected": "a maneuver is already in progress"}

        from_alt, from_plane = _parse(asset.true_regime)
        to_alt, to_plane = _parse(target_regime)

        alt_fuel, alt_turns = ALTITUDE_COST[f"{from_alt}-{to_alt}"]
        plane_fuel, plane_turns = PLANE_COST[from_alt][f"{from_plane}-{to_plane}"]

        if alt_fuel > 0 and plane_fuel > 0:
            fuel_cost = int((alt_fuel + plane_fuel) * (1 - COMBINED_MANEUVER_DISCOUNT))
            turns_required = max(alt_turns, plane_turns) + 1
        elif alt_fuel > 0:
            fuel_cost, turns_required = alt_fuel, alt_turns
        else:
            fuel_cost, turns_required = plane_fuel, plane_turns

        return {"turns_required": turns_required, "fuel_cost": fuel_cost, "rejected": None}

    def maneuver_complete(self, asset: Asset) -> bool:
        if asset.maneuver_state is None:
            return False
        if asset.maneuver_state.turns_remaining > 0:
            return False
        asset.true_regime = asset.maneuver_state.target_regime
        asset.maneuver_state = None
        return True

File: server/src/engine/Propagator.ts
/**
 * Propagator (IP-5010) — two-body Keplerian propagation (MSTR-001 C4 v0.3: no J2 term for v1),
 * discrete-regime classification (R-203's 9 labels), and maneuver planning/completion using
 * FS-104's Maneuver Cost Table (grounded in R-201). Sole module holding continuous orbital
 * state (GDS-03's isolation boundary) — every external read goes through currentRegime, never a
 * raw element.
 */
import type { Asset, OrbitalRegimeLabel } from '@owchess/shared';

type AltitudeBand = 'LEO' | 'MEO' | 'GEO';
type PlaneClass = 'EQUATORIAL' | 'PROGRADE' | 'POLAR';

const ALTITUDE_BANDS: AltitudeBand[] = ['LEO', 'MEO', 'GEO'];
const PLANE_CLASSES: PlaneClass[] = ['EQUATORIAL', 'PROGRADE', 'POLAR'];

/** Mean motion (degrees per turn-tick), a fixed per-altitude-band constant — faster in lower
 *  orbits, matching real orbital mechanics (shorter period at lower altitude). Internal
 *  bookkeeping only; never exposed past currentRegime. */
const MEAN_MOTION_DEG_PER_TICK: Record<AltitudeBand, number> = {
  LEO: 24, // fast — short real-world period (~90 min)
  MEO: 6, // GPS-like, ~12h period
  GEO: 1.5, // ~24h period
};

interface InternalOrbitalState {
  altitude: AltitudeBand;
  plane: PlaneClass;
  meanAnomalyDeg: number;
}

function parseRegime(label: OrbitalRegimeLabel): { altitude: AltitudeBand; plane: PlaneClass } {
  const [altitude, plane] = label.split('-') as [AltitudeBand, PlaneClass];
  return { altitude, plane };
}

function regimeLabel(altitude: AltitudeBand, plane: PlaneClass): OrbitalRegimeLabel {
  return `${altitude}-${plane}` as OrbitalRegimeLabel;
}

/** FS-104 Maneuver Cost Table §Altitude component. */
const ALTITUDE_COST: Record<string, { fuel: number; turns: number }> = {
  'LEO-LEO': { fuel: 0, turns: 0 },
  'MEO-MEO': { fuel: 0, turns: 0 },
  'GEO-GEO': { fuel: 0, turns: 0 },
  'LEO-MEO': { fuel: 3, turns: 2 },
  'MEO-LEO': { fuel: 3, turns: 2 },
  'MEO-GEO': { fuel: 1, turns: 3 },
  'GEO-MEO': { fuel: 1, turns: 3 },
  'LEO-GEO': { fuel: 4, turns: 4 },
  'GEO-LEO': { fuel: 4, turns: 4 },
};

/** FS-104 Maneuver Cost Table §Plane-class component, evaluated at the given (starting) altitude. */
const PLANE_COST: Record<AltitudeBand, Record<string, { fuel: number; turns: number }>> = {
  LEO: {
    'EQUATORIAL-EQUATORIAL': { fuel: 0, turns: 0 },
    'PROGRADE-PROGRADE': { fuel: 0, turns: 0 },
    'POLAR-POLAR': { fuel: 0, turns: 0 },
    'EQUATORIAL-PROGRADE': { fuel: 6, turns: 2 },
    'PROGRADE-EQUATORIAL': { fuel: 6, turns: 2 },
    'PROGRADE-POLAR': { fuel: 6, turns: 2 },
    'POLAR-PROGRADE': { fuel: 6, turns: 2 },
    'EQUATORIAL-POLAR': { fuel: 11, turns: 3 },
    'POLAR-EQUATORIAL': { fuel: 11, turns: 3 },
  },
  MEO: {
    'EQUATORIAL-EQUATORIAL': { fuel: 0, turns: 0 },
    'PROGRADE-PROGRADE': { fuel: 0, turns: 0 },
    'POLAR-POLAR': { fuel: 0, turns: 0 },
    'EQUATORIAL-PROGRADE': { fuel: 3, turns: 1 },
    'PROGRADE-EQUATORIAL': { fuel: 3, turns: 1 },
    'PROGRADE-POLAR': { fuel: 3, turns: 1 },
    'POLAR-PROGRADE': { fuel: 3, turns: 1 },
    'EQUATORIAL-POLAR': { fuel: 5, turns: 2 },
    'POLAR-EQUATORIAL': { fuel: 5, turns: 2 },
  },
  GEO: {
    'EQUATORIAL-EQUATORIAL': { fuel: 0, turns: 0 },
    'PROGRADE-PROGRADE': { fuel: 0, turns: 0 },
    'POLAR-POLAR': { fuel: 0, turns: 0 },
    'EQUATORIAL-PROGRADE': { fuel: 2, turns: 1 },
    'PROGRADE-EQUATORIAL': { fuel: 2, turns: 1 },
    'PROGRADE-POLAR': { fuel: 2, turns: 1 },
    'POLAR-PROGRADE': { fuel: 2, turns: 1 },
    'EQUATORIAL-POLAR': { fuel: 4, turns: 1 },
    'POLAR-EQUATORIAL': { fuel: 4, turns: 1 },
  },
};

/** Combined-maneuver discount (FS-104 §Maneuver Cost Table): both axes change in one call. */
const COMBINED_MANEUVER_DISCOUNT = 0.25;

export class Propagator {
  private states = new Map<string, InternalOrbitalState>();

  private stateFor(asset: Asset): InternalOrbitalState {
    let state = this.states.get(asset.assetId);
    if (!state) {
      const { altitude, plane } = parseRegime(asset.trueRegime);
      state = { altitude, plane, meanAnomalyDeg: 0 };
      this.states.set(asset.assetId, state);
    }
    return state;
  }

  /** Advances every tracked asset's true orbital position by one turn-tick (NFR-2100:
   *  deterministic given fixed prior state). A stable orbit's altitude/plane class do not change
   *  from mere propagation (real two-body mechanics) — only a completed maneuver changes them. */
  advance(assets: Asset[]): void {
    for (const asset of assets) {
      const state = this.stateFor(asset);
      state.meanAnomalyDeg = (state.meanAnomalyDeg + MEAN_MOTION_DEG_PER_TICK[state.altitude]) % 360;
    }
  }

  /** FR-5200: discrete presentation only — never exposes raw elements past this call's return. */
  currentRegime(asset: Asset): OrbitalRegimeLabel {
    const state = this.stateFor(asset);
    return regimeLabel(state.altitude, state.plane);
  }

  /**
   * FS-104's Maneuver Cost Table. BL-0014: rejects outright if a maneuver is already in progress
   * (GDS-07's single-field maneuverState schema — no queuing/replacement).
   */
  planManeuver(
    asset: Asset,
    targetRegime: OrbitalRegimeLabel
  ): { turnsRequired: number; fuelCost: number; rejected?: string } {
    if (asset.maneuverState !== null) {
      return { turnsRequired: 0, fuelCost: 0, rejected: 'a maneuver is already in progress (BL-0014)' };
    }
    const from = parseRegime(this.currentRegime(asset));
    const to = parseRegime(targetRegime);

    const altitude = ALTITUDE_COST[`${from.altitude}-${to.altitude}`];
    const plane = PLANE_COST[from.altitude][`${from.plane}-${to.plane}`];

    let fuelCost: number;
    let turnsRequired: number;
    if (altitude.fuel > 0 && plane.fuel > 0) {
      fuelCost = Math.floor((altitude.fuel + plane.fuel) * (1 - COMBINED_MANEUVER_DISCOUNT));
      turnsRequired = Math.max(altitude.turns, plane.turns) + 1;
    } else if (altitude.fuel > 0) {
      fuelCost = altitude.fuel;
      turnsRequired = altitude.turns;
    } else {
      fuelCost = plane.fuel;
      turnsRequired = plane.turns;
    }
    return { turnsRequired, fuelCost };
  }

  /**
   * True if an in-progress maneuver has completed this owner-turn-advance; finalizes trueRegime
   * and internal state, clearing maneuverState, when it has.
   */
  maneuverComplete(asset: Asset): boolean {
    if (asset.maneuverState === null) return false;
    if (asset.maneuverState.turnsRemaining > 0) return false;

    const target = parseRegime(asset.maneuverState.targetRegime);
    const state = this.stateFor(asset);
    state.altitude = target.altitude;
    state.plane = target.plane;
    asset.trueRegime = asset.maneuverState.targetRegime;
    asset.maneuverState = null;
    return true;
  }
}
